packages/next/src/server/lib/router-server.ts

```diff
@@ -207,39 +214,57 @@ export async function initialize(opts: {
   )
 
   // pre-initialize workers
-  await renderWorkers.app?.initialize(renderWorkerOpts)
-  await renderWorkers.pages?.initialize(renderWorkerOpts)
+  const initialized = {
+    app: await renderWorkers.app?.initialize(renderWorkerOpts),
+    pages: await renderWorkers.pages?.initialize(renderWorkerOpts),
+  }
 
   if (devInstance) {
     Object.assign(devInstance.renderWorkers, renderWorkers)
+
+    nextDeleteCacheCallbacks.push((filePaths: string[]) =>
+      Promise.all([
+        renderWorkers.pages?.deleteCache(filePaths),
+        renderWorkers.app?.deleteCache(filePaths),
+      ])
+    )
+    nextDeleteAppClientCacheCallbacks.push(() =>
+      Promise.all([
+        renderWorkers.pages?.deleteAppClientCache(),
+        renderWorkers.app?.deleteAppClientCache(),
+      ])
+    )
+    nextClearModuleContextCallbacks.push((targetPath: string) =>
+      Promise.all([
+        renderWorkers.pages?.clearModuleContext(targetPath),
+        renderWorkers.app?.clearModuleContext(targetPath),
+      ])
+    )
     ;(global as any)._nextDeleteCache = async (filePaths: string[]) => {
-      try {
-        await Promise.all([
-          renderWorkers.pages?.deleteCache(filePaths),
-          renderWorkers.app?.deleteCache(filePaths),
-        ])
-      } catch (err) {
-        console.error(err)
+      for (const cb of nextDeleteCacheCallbacks) {
```

Comment:
should those callbacks be parallelised?

Reply:
@ijjk it's unclear to me why these are arrays, seems they're all arrays with only one item? 🤔